File: node/streaming_speaker_diarization_speaker_id.js
const fs = require("fs");
const { SpeechClient } = require("@soniox/soniox-node");

/**
 * Set your Soniox Cloud API key:
 * from command line: export SONIOX_API_KEY=<YOUR-API-KEY>
 * or
 * pass config object: new SpeechClient({ api_key: "<YOUR-API-KEY>" })
 */
const speechClient = new SpeechClient();

(async function () {
  const onDataHandler = async (result) => {
    let speaker_num_to_name = {}
    for (const entry of result.speakers) {
      speaker_num_to_name[entry.speaker] = entry.name
    }

    const getName = (speaker) => {
      if (speaker in speaker_num_to_name) {
        return speaker_num_to_name[speaker]
      } else {
        return "unknown"
      }
    };

    console.log(result.words.map((word) =>
      `${word.text}/${word.speaker}(${getName(word.speaker)})`).join(" ")
    );
  };

  const onEndHandler = (error) => {
    if (error) {
      console.log(error);
    }
  };

  // transcribeStream returns object with ".writeAsync()" and ".end()" methods
  // use them to send data and end stream when done.
  const stream = speechClient.transcribeStream(
    {
      enable_streaming_speaker_diarization: true,
      enable_speaker_identification: true,
      cand_speaker_names: ["John", "Judy"],
      include_nonfinal: true
    },
    onDataHandler,
    onEndHandler
  );

  const CHUNK_SIZE = 1024;
  const readable = fs.createReadStream("../test_data/test_audio_sd.flac", {
    highWaterMark: CHUNK_SIZE,
  });

  // Simulate data streaming
  for await (const chunk of readable) {
    stream.writeAsync(chunk);
  }

  stream.end();
})();

File: python/streaming_speaker_diarization_speaker_id.py
from typing import Iterable
from soniox.transcribe_live import transcribe_stream
from soniox.speech_service import SpeechClient


def iter_audio() -> Iterable[bytes]:
    with open("../test_data/test_audio_sd.flac", "rb") as fh:
        while True:
            audio = fh.read(1024)
            if len(audio) == 0:
                break
            yield audio


def main():
    with SpeechClient() as client:
        for result in transcribe_stream(
            iter_audio(),
            client,
            model="en_v2_lowlatency",
            include_nonfinal=True,
            enable_streaming_speaker_diarization=True,
            enable_speaker_identification=True,
            cand_speaker_names=["John", "Judy"],
        ):
            speaker_num_to_name = {entry.speaker: entry.name for entry in result.speakers}

            def get_name(speaker):
                if speaker in speaker_num_to_name:
                    return speaker_num_to_name[speaker]
                else:
                    return "unknown"

            print(" ".join(f"'{w.text}'/{w.speaker}({get_name(w.speaker)})" for w in result.words))


if __name__ == "__main__":
    main()
